core/include/gap/core/optional.hpp

```diff
@@ -0,0 +1,225 @@
+// Copyright (c) 2022-present, Trail of Bits, Inc.
+
+#pragma once
+
+#include <concepts>
+#include <optional>
+#include <type_traits>
+
+#include <gap/core/concepts.hpp>
+#include <gap/core/pointer.hpp>
+
+namespace gap {
+
+    //
+    // optional concepts
+    //
+    template< typename T >
+    concept optional_like = requires (T a) {
+        { a.value() }     -> convertible_to< typename T::value_type >;
+        { a.has_value() } -> convertible_to< bool >;
+        { *a } -> convertible_to< typename T::value_type >;
```

Comment:
What about default constructibility? Is it not needed for `and_then` and others?

Reply:
true